cis194/week6/daniel-mc/Fibonacci.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -Wall #-}
+
+module Fibonacci
+  ( fib
+  , fibs1
+  , fibs2
+  , streamToList
+  , streamRepeat
+  , streamMap
+  , streamFromSeed
+  , nats
+  , ruler
+  , fibs3
+  , fibs4
+  )
+where
+
+-- Exercise 1
+
+fib :: Integer -> Integer
+fib 0 = 0
+fib 1 = 1
+fib n = fib (n - 1) + fib (n - 2)
+
+fibs1 :: [Integer]
+fibs1 = map fib [0 ..]
+
+
+-- Exercise 2
+iterate2 :: (a -> a -> a) -> a -> a -> [a]
+iterate2 f x y = x : y : iterate2 f x' (f y x') where x' = f x y
+
+fibs2 :: [Integer]
+fibs2 = iterate2 (+) 0 1
+
+-- Exercise 3
+data Stream a = Stream a (Stream a)
+
+streamToList :: Stream a -> [a]
+streamToList (Stream x xs) = x : streamToList xs
+
+instance Show a => Show (Stream a) where
+  show = show . take 20 . streamToList
+
+-- Exercise 4
+streamRepeat :: a -> Stream a
+streamRepeat x = Stream x (streamRepeat x)
```

Comment:
FYI - if you would like to have a point-free version here:

```suggestion
streamRepeat = Stream <*> streamRepeat
```

Reply:
Thanks, but I might leave it out for now since I know very little about things like Applicative and Functor.